### DISCUSSION TITLE
Python version

### DISCUSSION THREAD

**Discussion by brno32** [paustint/soql-parser-js#165] [Ideas]
I wrote https://github.com/Kicksaw-Consulting/python-soql-parser but it's not very good, has a lot limits, and the code sucks. I'm ashamed of it.

I'm looking for someone who can help me rewrite it, hopefully taking inspiration from this library. If a rewrite is not possible, is there a feasible way to transpile this lib to python code, or run it from within python using something like https://github.com/PiotrDabkowski/Js2Py?

Hoping to start a good discussion here! Thanks for putting this library out there.

**Comment by paustint** [comment]
I think your best bet might be to try to find an existing Antlr grammar for soql and use that for parsing, then you just need to build the logic to visit the code and generate a data structure of your choosing.

**Comment by brno32** [comment]
I'm currently using pyparsing but coming up against my intellectual limits when I try to expand what it can already do. I've only got it supporting fairly basic queries at the moment. Perhaps I should try another library.

Would you entertain using a tool like jsii for this library? https://github.com/aws/jsii/ It works well with cdk.

Since I ultimately want a soql-parser for my unit tests, speed isn't a concern to me.

**Comment by paustint** [comment]
@brno32 - I take a few minutes looking into jsii and I think it might take some working with to figure out how to make this work. It looks like the required typescript options are a bit stricter than I have implemented, so it will require some minor tweaks to accommodate.

I will try to spend more time investigating to see if this is a valid option and hopefully it is.

**Comment by brno32** [comment]
Great to hear. Happy to help if possible!

**Comment by paustint** [comment]
@brno32 - jsii adds too many restrictions to what is allowed in Typescript and it would be a significant amount of effort and refactoring just to get the codebase compatible.

I did just release a new version that includes a CLI, so that could be a potentially option for you - but it will require that you have the node runtime installed, then you could use something like [toolshed_shell](https://naked.readthedocs.io/toolshed_shell.html) to run the cli. Check out the readme for some examples.

Example (might be outdated): http://sweetme.at/2014/02/17/a-simple-approach-to-execute-a-node.js-script-from-python/

I hope this is a viable option based on your environment. The other option would be to spin up a node API server and validate over http, but that would be really slow and could be error prone - but would be fairly trivial to setup as long as you had somewhere to host it.

**Reply by brno32** [reply]
This could be a viable option! Since I'd need the node runtime available to run jsii anyway, I was already mulling over this. I think the cli would be a good solution.

Thanks for investigating!

**Reply by paustint** [reply]
Awesome! I am not sure how familiar with node you are, but in case you are not familiar, here is how you get started.

After installing node, you have two options:

install soql-parser js globally `npm install soql-parser-js -g`
Then invoke with `soql-parser-js parse "SELECT Id FROM Account"`

**OR**

Run on-demand, may not be great in your testing scenario since it makes network calls on each invocation (not sure how caching works)
`npx soql-parser-js parse "SELECT Id FROM Account"`


It might make sense to Dockerize your testing runs so that you can have a reproducible and portable environment.
